### ISSUE TITLE
problem with polymorphic 'this' type in bound generics

### ISSUE THREAD

**Issue by Artazor** [Microsoft/TypeScript#6621]
I've defined two interfaces: ([this is synthetic example](http://www.typescriptlang.org/Playground#src=interface%20BaseNode%20%7B%20next%3F%3A%20this%3B%20%7D%0Ainterface%20SomeNode%20extends%20BaseNode%20%7B%20a%3A%20number%3B%20%7D%0A%0Avar%20node%3A%20SomeNode%20%3D%20%7Ba%3A%201%2C%20next%3A%20%7Ba%3A%202%2C%20next%3A%20%7Ba%3A%203%7D%7D%7D%0Avar%20arr%20%3D%20%5B%5D%2C%20i%20%3D%200%3B%0Awhile%20(node)%20%7B%0A%09arr%5Bi%5D%20%3D%20node%3B%0A%09i%2B%2B%3B%0A%09node%20%3D%20node.next%3B%20%2F%2F%20As%20expected%20(no%20errors)%20%0A%7D%0Aconsole.log(arr)%3B%0A%0Afunction%20toArray%3CT%20extends%20BaseNode%3E(node%3A%20T)%3A%20Array%3CT%3E%20%7B%0A%09var%20arr%20%3D%20%5B%5D%2C%20i%20%3D%200%3B%0A%09while%20(node)%20%7B%0A%09%09arr%5Bi%5D%20%3D%20node%3B%0A%09%09i%2B%2B%3B%0A%09%09node%20%3D%20node.next%3B%20%2F%2F%20Error%3A%20Type%20'BaseNode'%20is%20not%20assignable%20to%20type%20'T'%0A%09%7D%0A%09return%20arr%3B%0A%7D))

```TypeScript
interface BaseNode { next?: this; }
interface SomeNode extends BaseNode { a: number; }
```
then if i write something like
```TypeScript
var node: SomeNode = {a: 1, next: {a: 2, next: {a: 3}}};
var arr = [], i = 0;
while (node) {
	arr[i] = node;
	i++;
	node = node.next; // As expected (no errors) 
}
console.log(arr)
```
then all compiles like a charm. On the other side when I'm trying to write a function that does the same and has the correct signature, I'm getting an error:
```TypeScript
function toArray<T extends BaseNode>(node: T): Array<T> {
	var arr = [], i = 0;
	while (node) {
		arr[i] = node;
		i++;
		node = node.next; // Error: Type 'BaseNode' is not assignable to type 'T'
	}
	return arr;
}
```
Should it compile without any errors?

**Comment by falsandtru** [comment]
It may not depend polymorphic `this` type.

```ts
function f<T extends number>(n: T) {
	n = 0; // base to derived, error
	var m: number = n; // derived to base, ok
}
```

**Comment by ahejlsberg** [comment]
Duplicate of #5525. It was fixed in #5552 and your example works with current compiler in master.
